### PR TITLE
Restore object implementation

### LINKED ISSUE #2087: UX: Easy way to restore after spying on all methods of an object
In #2027 we gained the ability to spy on all methods of an object. However the only way to restore the object is to use a sandbox or visit every method of the object. It would be a convenient and a better user experience if `spyOnEntireObj` returned a way to revert everything back to normal.

### PR DESCRIPTION
**Pull request by fearphage**
Took a quick stab at this one, I probably forgot a lot of corner cases. Feel free to help me flesh this out.

Fixes #2087

### CODE REVIEW

File: test/restore-object.js [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        assert.isFunction(restoreObject);
+    });
+
+    it("works on empty objects", function() {
+        refute.exception(function() {
+            restoreObject({});
+        });
+    });
+
+    it("quietly ignores falsey input", function() {
```
Comment (mroderick) [inline test/restore-object.js]:
```suggestion
    it("quietly ignores falsy input", function() {
```

File: lib/sinon/restore-object.js [hunk 2]
```diff
@@ -0,0 +1,22 @@
+"use strict";
+
+var getPropertyDescriptor = require("./util/core/get-property-descriptor");
+var walk = require("./util/core/walk");
+
+function restoreObject(object) {
+    walk(object || {}, function(prop, propOwner) {
+        if (
```
Comment (mroderick) [inline lib/sinon/restore-object.js]: You can reduce the indentation by using the return early pattern here.

This is very much a matter of style/preference, feel free to ignore.

Comment (fearphage) [reply lib/sinon/restore-object.js]: This feels better to me. If there was more than one line of code inside the `if`, I would be more compelled to switch this around. Also this was copied from one of the other files so it follows the same pattern.

**Comment by fearphage** [comment]
> You might want to add some documentation for it

I added a ticket for the docs. #2089

File: test/restore-object.js [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+var restoreObject = require("../lib/sinon/restore-object.js");
+var assert = referee.assert;
+var refute = referee.refute;
+
+describe("restore-object", function() {
+    it("is defined", function() {
+        assert.isFunction(restoreObject);
+    });
+
+    it("works on empty objects", function() {
```
Comment (mantoni) [inline test/restore-object.js]: Why should it? Sinon usually throws a meaninful exception in cases where passed arguments are not "Sinon objects". This helps finding issues while writing or refactoring test cases. Think of passing a non-fake object to `sinon.assert.called(xyz)`, for example. In case of this method I would prefer having an exception thrown if nothing is being restored. Other thoughts?

File: test/restore-object.js [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        refute.exception(function() {
+            restoreObject({});
+        });
+    });
+
+    it("quietly ignores falsy input", function() {
+        refute.exception(function() {
+            restoreObject(false);
```
Comment (mantoni) [inline test/restore-object.js]: Same as comment above. This should throw a meaningful exception. The user is doing something dodgy. The argument is clearly not a result of calling `sinon.stub(...)`.

File: test/restore-object.js [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            restoreObject(null);
+        });
+    });
+
+    it("works with no spies or stubs", function() {
+        refute.exception(function() {
```
Comment (mantoni) [inline test/restore-object.js]: Should throw in my opinion. Again, not the result of calling `sinon.stub` being passed.

File: test/restore-object.js (hunk 3 above)
Comment (fearphage) [reply test/restore-object.js]: @mantoni This is the way both `spyOnEntireObject` and `stubEntireObject` are defined. Should those change as well?

```shell
➜  sinon git:(restore-object) ✗ ag -Q 'walk(object || {}' lib
lib/sinon/stub-entire-object.js
7:    walk(object || {}, function(prop, propOwner) {

lib/sinon/spy-entire-object.js
7:    walk(object || {}, function(prop, propOwner) {

lib/sinon/restore-object.js
7:    walk(object || {}, function(prop, propOwner) {
```

File: test/restore-object.js (hunk 4 above)
Comment (fearphage) [reply test/restore-object.js]: Same as my comment above.

File: test/restore-object.js (hunk 3 above)
Comment (mantoni) [reply test/restore-object.js]: I lean towards failing in the new API. We can make it consistent now, or later as a breaking change. I have no specific opinion here. Do @fatso83 and @mroderick have additional arguments?

Comment (fearphage) [reply test/restore-object.js]: What does failing do/mean? 

From my perspective, the user asks to restore all mocks, spies, and stubs. We return an object with 0 mocks, spies, and stubs.

The error case makes sense for 1-to-1 mappings of "do X to this spy/stub". It feels a bit more ambiguous here though.

Comment (fatso83) [reply test/restore-object.js]: Break things! Yeah! I always like to land on the strict side of things, so :+1: to not allowing stubbing or restoring `undefined`.

Comment (fatso83) [reply test/restore-object.js]: But ... we could do it in a separate PR just to not bloat this one.

Comment (fearphage) [reply test/restore-object.js]: I like the idea of doing them outside of this fix. 

We should make them all congruent. That sounds like a very breaking change.

Comment (mantoni) [reply test/restore-object.js]: It makes sense to keep the implementations symetric for now add strictness later. I have created #2090 summarizing what was discussed here. 👍

Comment (fearphage) [reply test/restore-object.js]: I started working on #2090 in a branch(based off of this one).

**Comment by fearphage** [comment]
@mroderick If you're happy with this, I'll merge and move on to making this stricter in the next PR.

**Comment by fearphage** [comment]
I'm addressing issues addressed here in #2092.
